serial.go

```diff
@@ -19,27 +19,44 @@ package main
 
 import (
 	"encoding/json"
+	"fmt"
 	"slices"
-	"strconv"
 	"strings"
 	"sync"
 	"time"
 
+	"github.com/arduino/arduino-create-agent/tools"
 	discovery "github.com/arduino/pluggable-discovery-protocol-handler/v2"
 	"github.com/sirupsen/logrus"
 )
 
 type serialhub struct {
 	// Opened serial ports.
 	ports map[*serport]bool
+	mu    sync.Mutex
 
-	mu sync.Mutex
+	OnRegister   func(port *serport)
+	OnUnregister func(port *serport)
 }
 
-// SerialPortList is the serial port list
-type SerialPortList struct {
+func newSerialHub() *serialhub {
+	return &serialhub{
+		ports: make(map[*serport]bool),
+	}
+}
+
+type serialPortList struct {
+	tools *tools.Tools
+
 	Ports     []*SpPortItem
 	portsLock sync.Mutex
+
+	OnList func([]byte) `json:"-"`
+	OnErr  func(string) `json:"-"`
+}
+
+func newSerialPortList(tools *tools.Tools) *serialPortList {
```

Comment:
to make this clearer I would create a separate package for this we can put it in the main folder for now but maybe there is a bettere place in the future